lib/builder.js

```diff
@@ -219,6 +220,207 @@ Builder.prototype.traceModule = function(moduleName) {
+  }
+
+  if(typeof optimizationFunction !== 'function') {
+    throw "optimizationFunction must be a function.";
+  }
+
+  var self = this,
+      bundleNames = [],
+      bundleNameMap = {};
+
+  return Promise.all(Object.keys(entryPoints).map(function(bundleNameKey, bundleNameKeyIndex) {
+    // This block traces each entry point and ensures it has a name, even if default
+    bundleNames[bundleNameKeyIndex] = (entryPoints instanceof Array) ? 'bundle'+bundleNameKey : bundleNameKey;
+    bundleNameMap[bundleNames[bundleNameKeyIndex]] = bundleNameKey; // bundleNameMap helps map the (possible generated) bundle name back to the entryPoints index
+    return self.trace(entryPoints[bundleNameKey], opts.config);
+  })).
+  then(function(traces) {
+    // Take the traces and places them in an object indexed by bundle names
+    var namedTraces = traces.reduce(function(previous, current, index) {
+      previous[bundleNames[index]] = current;
+      return previous;
+    }, {});
+
+    opts = extend({ bundleNameMap: bundleNameMap }, opts);
+
+    return optimizationFunction(entryPoints, namedTraces, opts); // run custom optimization function
+  }).
+  then(function(bundleData) {
+    var output = {
+          bundles: [],
+          config: bundleData.config
+        },
+        outPath;
+
+    // Super-specific validation on returned data as that user-provided function could do anything!
+    Builder.validateOptimizedBundleData(bundleData);
+
+    if(opts.outPath) {
+      outPath = path.resolve(opts.outPath);
+    }
+
+    return Promise.all(Object.keys(bundleData.bundles).map(function(bundleIndex) {
+      // For each bundle returned from the optimization function...
+      // ...create an object to represent it in the response from optimizeBuild
+      var bundleName = bundleData.bundles[bundleIndex].name,
+          outputBundle = {
+            name: bundleName,
+            entryPoints: bundleData.bundles[bundleIndex].entryPoints,
+            modules: bundleData.bundles[bundleIndex].modules
+          };
+
+      output.bundles.push(outputBundle);
+
+      // ...build the code and either write the file (automatically as part of buildTree)...
+      return self.buildTree(bundleData.bundles[bundleIndex].tree, outPath ? outPath+path.sep+bundleName+'.js' : null, {
+        minify: opts.minify === undefined ? true : !!opts.minify,
+        sourceMaps: opts.sourceMaps === undefined ? true : !!opts.sourceMaps,
+        mangle: opts.mangle === undefined ? true : !!opts.mangle
+      },
+      function(err) {
+        throw "optimizeBuild: error creating bundle `"+bundleName+"`: "+err;
+      }).
+      then(function(result) {
+        // ...or save the built-tree source as a property on the bundle representation.
+        if(!outPath) {
+          outputBundle.source = result;
+        }
+      });
+    })).
+    then(function() {
+      var promise = new Promise(function(resolve, reject) {
```

Comment:
As mentioned, it would be nice to just generate `output.bundles` here.

I think `output.depCache` is a roughly equivalent output anyway, or is it different?

This would allow the user to put the config with the entry points as needed. For example, you probably want to only include the bundles config that applies for specific rendered pages.